tools/qlog/src/lib.rs

```diff
@@ -258,6 +298,264 @@ impl Default for Qlog {
+    log_finished: bool,
+    first_event: bool,
+    first_frame: bool,
+    writing_event: bool,
+    writing_frames: bool,
+}
+
+impl QlogStreamer {
+    /// Creates a QlogStreamer object.
+    ///
+    /// It owns a `Qlog` object that contains the provided `Trace`, which must
+    /// have the following ordered-set of names EventFields:
+    ///
+    /// ["relative_time", "category", "event".to_string(), "data"]
+    ///
+    /// All serialization will be written to the provided `Write`.
+    pub fn new(
+        qlog_version: String, title: Option<String>, description: Option<String>,
+        summary: Option<String>, start_time: std::time::Instant, trace: Trace,
+        writer: Box<dyn std::io::Write>,
+    ) -> Self {
+        let qlog = Qlog {
+            qlog_version,
+            title,
+            description,
+            summary,
+            traces: vec![trace],
+        };
+
+        QlogStreamer {
+            start_time,
+            writer,
+            qlog,
+            first_event: true,
+            first_frame: false,
+            writing_event: false,
+            writing_frames: false,
+            log_started: false,
+            log_finished: false,
+        }
+    }
+
+    /// Starts qlog streaming serialization.
+    ///
+    /// This writes out the JSON-serialized form of all information up to qlog
+    /// `Trace`'s array of `EventField`s. EventFields are separately appended
+    /// using functions that accept and `event::Event`.
+    pub fn write_start(&mut self) -> Result<()> {
+        // We can only start the log once.
+        if self.log_started {
+            return Err(Error::Done);
+        }
+
+        // A qlog contains a trace holding a vector of events that we want to
+        // serialize in a streaming manner. So at the start of serialization,
+        // take off all closing delimiters, and leave us in a state to accept
+        // new events.
+        // TODO: this unfortunately removes any events already existing in the
+        // trace, which might be undesirable.
+        match serde_json::to_string(&self.qlog) {
+            Ok(mut out) => {
+                if let Some(index) = out.find("\"events\":[") {
+                    out.truncate(index + "\"events\":[".len());
+                }
+
+                self.writer.as_mut().write_all(out.as_bytes())?;
+
+                self.log_started = true;
+            },
+
+            _ => return Err(Error::Done),
+        }
+
+        Ok(())
+    }
+
+    /// Finishes qlog streaming serialization.
+    ///
+    /// The JSON-serialized output has remaining close delimiters added.
+    /// After this is called, no more serialization will occur.
+    pub fn log_finish(&mut self) -> Result<()> {
+        if !self.log_started || self.log_finished {
+            return Err(Error::InvalidState);
+        }
+
+        self.writer.as_mut().write_all(b"]}]}")?;
+
+        self.log_finished = true;
+
+        self.writer.as_mut().flush()?;
```

Comment:
Seems like this could impact performance, do we really need to do this?

Reply:
I only added this recently because I noticed a problem when that testing the C API. In that test program, it opened the file but no data was written, adding flush here resolved the problem. But actually I now realise that this indicates the log is getting buffered somewhere. This doesn't happen in rust natively. Any ideas? 

`log_finish()` can only be called once so it didn't seem to be too much overhead in ensuring all data was flushed at the end of a connection.

Reply:
I think buffers will be flushed depending on the timing when this is out of scope? Or it may help if you explicitly close the writer (drop?) here.